Add random compositional fields

alarshi opened the pull request:
This PR addresses part of issue #531 to get random compositional values. Here is the output of this plugin with a continental plate:

![image](https://github.com/GeodynamicWorldBuilder/WorldBuilder/assets/19296817/8cf98367-ae57-4953-be5b-9bcb9d8eb26f)
This is the result for user specified bounds in the first compositional field: [5, 10]

![image](https://github.com/GeodynamicWorldBuilder/WorldBuilder/assets/19296817/0a429239-8f33-4e9a-822a-17b82fd5d370)
This is the result from using the default `random` model.

MFraters commented on source/world_builder/features/continental_plate_models/composition/random.cc:
```diff
@@ -0,0 +1,138 @@
+                    {
+                      if (compositions[i] == composition_number)
+                        {
+                          const double composition_ = ((double)rand()) / RAND_MAX * (bound_values[1] - bound_values[0]) + bound_values[0];
```
This works, but the following would be more in line with how the rest of the code generates random numbers.

```suggestion
                          std::uniform_real_distribution<> dist(min_value,max_value);
                          const double composition_ = dist(world->get_random_number_engine());
```

MFraters commented on source/world_builder/features/continental_plate_models/composition/random.cc:
```diff
@@ -0,0 +1,138 @@
+          prm.declare_entry("compositions", Types::Array(Types::UnsignedInt(),0),
+                            "A list with the labels of the composition which are present there.");
+
+          prm.declare_entry("bound values", Types::Array(Types::Double(1), (0, 1)),
```
Can you turn this into a `Max value` and a `Min value` or something similar? I think it is better to have them as separate entries.

MFraters commented on source/world_builder/features/continental_plate_models/composition/random.cc:
```diff
@@ -0,0 +1,138 @@
+          max_depth = max_depth_surface.maximum;
+
+          compositions = prm.get_vector<unsigned int>("compositions");
+          bound_values = prm.get_vector<double>("bound values");
```
Can you split it here as well?

MFraters reviewed (commented):
Looks great, thanks for making this! I have a few small comments, but otherwise I think it is good to go.

alarshi commented:
Thank you for the review! I addressed the comments in the latest commit. 
I was also wondering if the test case makes sense or not. Since, this is a random distribution, the output would be different every time we run it or do I not understand this correctly?

MFraters commented:
> I was also wondering if the test case makes sense or not. Since, this is a random distribution, the output would be different every time we run it or do I not understand this correctly?

The random numbers are depended on the seed. As long as the seed is the same, the same sequence of random numbers will be generated. If you update your test results, the tester should pass.

alarshi commented:
Got it! Thanks for the explanation.

MFraters commented:
hmm, strange. Did you rebase on the newest master when updating the test results?

alarshi commented:
I rebased the branch but the tests are still failing. I don't understand why. :(

alarshi commented:
I checked that each run is gives me a different result, which would explain why the test is failing.

tjhei commented:
The problem is that gwb-grid now uses multiple threads by default. This means that the points will not be evaluated in the same, deterministic order. I am not even sure that the random engine is thread-safe.
Making tests deterministic is not too difficult, by limiting all (or at least this) test to run with -j 1. If you want it to be deterministic for >1 thread, we would need to duplicate each random engine into a thread-local variable. This would make it deterministic as long as you run with the same number of threads but it would still depend on how many threads you run with. I am not sure if this is good enough...

MFraters commented:
ah, that make sense. I was confused, because we currently already use random number generation in the code, but this is for the grains, which is only implemented for the for gwb-dat, which doesn't use threads.

MFraters commented:
In any case, we should probably fix the number of threads for the tester.

MFraters commented:
hmm, getting a thread rank is not as easy as getting a mpi rank. The thread id seems to be an unpredictable number/hash (I think). We do actually have a unique rank number when creating them, but we would need to pass those id's manually somehow.

Looking at it, the easiest solution may be to add a constructor parameter to the world so that you can pass a thread rank. That rank is then use in some way to set the seed of the world. In gwb-grid, we then just create a separate world for each thread, and in the parallel for loop we call the properties function on the appropriate world for each thread. This would make it deterministic again and would require minimal changes I think.

MFraters commented:
@alarshi #667 is merged now, so if you rebase you should be able to fix this problem (at least updating the test results on my systems seems to work now). Also, it looks like you will need to include a header (I think the <random> header?) to make the non-unity build work.

MFraters reviewed (approved):
I am not sure why it didn't work for you, but it seems to work for me :s
